doc/specs/#147 - Support different release channels [released, beta, alpha].md

```diff
@@ -0,0 +1,259 @@
+---
+author: Kaleb Luedtke - Trenly <trenlymc@gmail.com>
+created on: 2021-11-18
+last updated: 2021-11-20
+issue id: 147
+---
+
+# Release Channels
+
+For [#147](https://github.com/microsoft/winget-cli/issues/147)
+
+## Abstract
+
+Several packages have different channels for releasing different versions of their software. The Windows Package Manager should be able to support packages of the same Package Identifier where the versions come from different release channels. The Windows Package Manager should be also able to inform a user about which channels are available.
+
+## Inspiration
+
+Some other package managers have this capability. In the current state, a new package identifier must be created for each release channel which can cause conflicts when the versions are *not* side-by-side compatible.
+
+## Solution Design
+
+### Manifest Structure
+The Windows Package Manager manifest v1.1.0 schema has provided a key for declaring which channel an installer belongs to. However, this has the potential to make the installer manifests extremely complex when there are multiple release channels with multiple installers each.
+> The `Channel` key should be removed from the manifest v1.1.0 schema.
+
+Instead, the use of multiple installer manifest files will be used to define and manage the various channels available for a package in a way similar to locales. This will ensure that the client remains backwards compatible with old manifest versions, and will make the creation and maintenance of the various channels easier for contributors in the community repository. This will require a key be added to the  manifest schema. Channels should not be compatible with singleton manifests.
+> The `DefaultChannel` key should be added to the manifest v1.1.0 schema for version manifests
+
+An example of the manifest file structure, with installer channel manifests:
+```raw
+/manifests/g/Google/Chrome/1.0
+  | Google.Chrome.installer.beta.yaml
+  | Google.Chrome.installer.canary.yaml
+  | Google.Crome.installer.dev.yaml
+  | Google.Chrome.installer.Stable.yaml
+  | Google.Chrome.locale.en-US.yaml
+  | Google.Chrome.locale.nb-NO.yaml
+  | Google.Chrome.yaml
+/manifests/g/Google/Chrome/2.0
+  | Google.Chrome.installer.canary.yaml
+  | Google.Crome.installer.dev.yaml
+  | Google.Chrome.locale.en-US.yaml
+  | Google.Chrome.locale.nb-NO.yaml
+  | Google.Chrome.yaml
```

Comment:
The problem I have with this is that it relies on the folder/file naming setup of the community repo, which the client is not aware of (since it is designed to work with any source). How would channels be expressed without a schema field in REST sources, or other pre-indexed sources?

I think we keep the channel field in the schema, and add the `DefaultChannel` to the version schema. The channel can be given per installer entry, which I'm aware will get unwieldy (but hopefully we don't have to use it often enough for it to be a problem).

Reply:
If I’m not mistaken, for other sources, the channel would still be included in the package schema. Since the package schema is different than the manifest schema, I believe this would still be supported?

Either that, or I would suggest having it both ways. That way the community repo and other sources  could still be kept relatively organized by taking advantage of splitting the manifests, but other sources which aren’t able to could still support channels

Reply:
The original thought (pre multi-file manifests) was to have a structure like:
```
publisher\package\version
publisher\package\non-default-channel\version
```
With `Channel` being in the installer schema, your design can work too.  I think it might be a bit more confusing though, as it is harder to see what is in each channel at a glance.

Reply:
> The original thought (pre multi-file manifests) was to have a structure like:
> 
> ```
> publisher\package\version
> publisher\package\non-default-channel\version
> ```
>  With Channel being in the installer schema, your design can work too. I think it might be a bit more confusing though, as it is harder to see what is in each channel at a glance.

That is the way it works currently, but based on the validation over at winget-pkgs, these would be completely separate package Identifiers. If I’m not mistaken, the whole point of channels is to combine package identifiers to be better representative of how packages are published, since some publishers do not allow side-by-side compatibility for different channels and so a user may not realize that Example.Package would be overwritten if they install Example.Package.Dev

Reply:
We would update winget-pkgs validation to handle non-empty channels in this way so that both would still be `Example.Package`.

Reply:
That would work too. If that is the preferred approach, I will update the spec. Just so long as we capture the work that is needed